### PR TITLE
Fixes #28516 - support all API operations

### PR DESCRIPTION
**Pull request by Ron-Lavi**
support also post, put, patch and delete requests in the API middleware

### CODE REVIEW

**Comment by laviro** [comment]
some notes:

- Adding the `config` key to API actions will provide all the configuration for the API request,
- thinking to change the action type from `API_OPERATIONS.[METHOD]` to `API_REQUEST`
and set the `method` in the `config` section
- the `url` key could still be set outside of the `config` section
- `config` will match `axios` configuration API.

**Comment by Ron-Lavi** [comment]
only the last commit is relevant, blocked by the work in #7222

**Comment by MariaAga** [comment]
Can you add  to the documentation the new info?

**Comment by mmoll** [comment]
@laviro please rebase :)

File: webpack/assets/javascripts/react_app/redux/API/__tests__/APIHelpers.test.js
Comment (mmoll) [inline webpack/assets/javascripts/react_app/redux/API/__tests__/APIHelpers.test.js]: commented code block? :)

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/__tests__/APIHelpers.test.js]: thanks, removed that

File: webpack/assets/javascripts/react_app/redux/API/APIRequest.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/redux/API/APIRequest.js]: Does it need to await here if there is an `await` in `getApiResponse`?

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIRequest.js]: thanks, this `await` should remain in here,
although it seem the the `await` in the helpers could be removed.

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIActions.js]: `apiAction()` looks a bit redundant to me.
doesn't it readable enough?
```js
export const APIActions = {
  get: payload => ({ GET, payload })
 ...
}

File: webpack/assets/javascripts/react_app/components/FactCharts/FactChartActions.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/FactCharts/FactChartActions.js]: does it make sense to import `get`, `post`, etc directly?

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIHelpers.js]: instead of iterating over a reversed object, you can just do a string manipulation thanks to our `type` convention:
```js
export const getApiMethod = type => type.substring(4).toLower() // 'API_GET' -> 'get'
```
covnetions over configurations :)

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/redux/API/APIHelpers.js]: how about returning it instead of storing it? it saves the need of a `break`
```js
return API[method](url, headers, params);
```
BTW, is there an option to use a general axios object instead of a switch case?
```js
axios({
  method,
  url,
  data: params
  headers,
  // etc
});

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @laviro ! 
few comments :)

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIActions.js]: I use it mostly for the method documentation,
to state what should the payload contain.
should we add the jsDoc over all of the methods?

File: webpack/assets/javascripts/react_app/components/FactCharts/FactChartActions.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/FactCharts/FactChartActions.js]: I wanted to distinguish between the `get` from the old `API.js`
and the action creator from `APIActions.js`.

another issue is that 'delete' is a reserved word in JS and you can't import it

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIHelpers.js]: that's awesome ! :)

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIHelpers.js]: that's a really good idea,
I was also investigated it in this PR,
though there are some tests mocks to the 'API.js' file that fails:

Selection_256

imo it would make more sense to open a new PR for it.

**Comment by Ron-Lavi** [comment]
ready for another review

**Review by amirfefer — CHANGES_REQUESTED** [review]
There's a runtime error after a rebase, probably due to a conflct with https://github.com/theforeman/foreman/pull/7121

```
Uncaught TypeError: (0 , _API.get) is not a function
    at getNotifications (index.js:21)
    at startNotificationsPolling (index.js:24)
    at redux.js:475
    at notificationContainer.componentDidMount (index.js:24)
    at commitLifeCycles (react-dom.development.js:20049)
    at commitLayoutEffects (react-dom.development.js:22813)
    at HTMLUnknownElement.callCallback (react-dom.development.js:347)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:397)
    at invokeGuardedCallback (react-dom.development.js:454)
    at commitRootImpl (react-dom.development.js:22585)
```

**Comment by Ron-Lavi** [comment]
Thanks @amirfefer, rebased

**Comment by Ron-Lavi** [comment]
rebased, ready for another review

**Review by amirfefer — CHANGES_REQUESTED** [review]
@laviro thanks! could you please use this implementation for post requests in [forms](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/redux/actions/common/forms.js#L54)? 
At least we will have one post request using this new middleware and it also verifies that it works as intended

**Comment by Ron-Lavi** [comment]
Thanks @amirfefer, rebased !
That implementation in forms seem to be a bit difficult to replace
since it is actually waiting for the API to return some data and trigger other actions based on it,
while the middleware update the state.. I will investigate more on it soon

**Comment by Ron-Lavi** [comment]
@amirfefer after looking more into replacing forms API call, we could:
- Add an interval to check when the state update (when status is not PENDING or so),
and than keep with the other dispatches there
- We could move that code to the API middleware to check if this action is FORM_SUBMIT

those two options feels like an hack to me,
I think the best option would be to redesign a bit the forms infrastructure around the API middleware in some better way, what are your thoughts @xprazak2 ?

**Comment by xprazak2** [comment]
I am not opposed to changes that would allow forms to leverage the API middleware (can be a follow up), we need a way to inform the form about the result of the call as already mentioned.

**Comment by Ron-Lavi** [comment]
as discussed offline, we could use the callbacks from #7337 to solve forms issue in this PR

**Comment by adamruzicka** [comment]
@amirfefer could you take a look?

**Review by amirfefer — DISMISSED** [review]
@laviro Thanks, 
Can you please rebase before merging?
looks like there are a few conflicts (BookmarksActions.js, statistics)

**Comment by Ron-Lavi** [comment]
Thanks @amirfefer, should we first merge #7337 as suggested in the [comment](https://github.com/theforeman/foreman/pull/7261#issuecomment-602538318) ? to solve also the forms example

**Comment by Ron-Lavi** [comment]
yay !! #7337 got merge,
so now I rebased and using the API middleware in forms,
supporting all of the API operations.

@amirfefer @xprazak2 @MariaAga I would glad to hear your thoughts about it.

File: webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/integration.test.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/integration.test.js:83]: Shouldn't the test check for something after the submit button is clicked? :)

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/redux/API/APIActions.js]: This is a bit confusing, can you change to `payload - payload` or `payload: payload`?

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/redux/API/APIHelpers.js:4]: This feels a bit hacky, it can be `split('API_')[1].toLowerCase()` so it will be more clear
or even `Object.keys(API_OPERATIONS).find(key => object[key] === type).toLowerCase();`

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/redux/API/APIActions.js:33]: Should also be a const like the rest maybe with a different name like `apiDelete`?

File: webpack/assets/javascripts/react_app/redux/API/__tests__/APIRequest.test.js
Comment (MariaAga) [inline webpack/assets/javascripts/react_app/redux/API/__tests__/APIRequest.test.js:25]: Can you send the new value in the test? `{type: ..., payload: {...action}`

**Review by MariaAga — COMMENTED** [review]
Looks and works great, left some small comments

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIHelpers.js:4]: it was requested to change by @amirfefer in: https://github.com/theforeman/foreman/pull/7261#discussion_r366440995

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIActions.js:33]: it's possible, though it's not an API request type..
I would prefer if someone uses it as:
```js
import { APIActions } ...
APIActions.delete()
```

File: webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/integration.test.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/integration.test.js:83]: Added integration snapshot that shows the whole process of the actions with the response and the success toast in the end

**Review by MariaAga — DISMISSED** [review]

**Comment by Ron-Lavi** [comment]
@ezr-ondrej @amirfefer @jeremylenz can you have another look? thanks :)

File: webpack/assets/javascripts/react_app/redux/API/APIActions.js
Comment (jeremylenz) [reply webpack/assets/javascripts/react_app/redux/API/APIActions.js:33]: I checked the [MDN docs here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Keywords) (under "Reserved word usage") and it looks like the way you've done it should be OK, since `delete` is being used as an "IdentifierName" and not an "Identifier."

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/redux/API/APIHelpers.js:14]: Would be nice as a follow-up PR to refactor `redux/API/API.js` so that all its methods take arguments in the same order.  Or even better, have the methods take a single object that can be destructured, like this one does.

File: webpack/assets/javascripts/react_app/redux/API/__tests__/APIMiddleware.test.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/redux/API/__tests__/APIMiddleware.test.js]:
```suggestion
  it('should not call apiRequest if action type is irrelevant', async () => {
```

File: webpack/assets/javascripts/react_app/redux/API/__tests__/APIMiddleware.test.js
Comment (jeremylenz) [inline webpack/assets/javascripts/react_app/redux/API/__tests__/APIMiddleware.test.js]:
```suggestion
  it('should call apiRequest when action type is relevant', async () => {
```

**Review by jeremylenz — COMMENTED** [review]
Thanks @laviro, I didn't retest but code is looking good!  Just a couple nits & comments below.

After this is merged I'm excited to start moving away from using thunks :)

File: webpack/assets/javascripts/react_app/redux/API/APIHelpers.js
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/API/APIHelpers.js:14]: yeah, I would even like to use `axios` directly from here

File: webpack/stories/docs/api-middleware.stories.mdx
Comment (ezr-ondrej) [inline webpack/stories/docs/api-middleware.stories.mdx]: Why the indentation change?

**Review by ezr-ondrej — COMMENTED** [review]
Overall looks great and it's much needed. I won't have time for a full review probably, but I can test it out once you feel it's ready.

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/redux/actions/common/forms.js:67]: @xprazak2 should we still dispatch this action after the API call secceed?
or maybe we can listen to the API middleware triggered actions, e.g:
`item_FORM_SUBMITTED_REQUEST`
`item_FORM_SUBMITTED_SUCCESS`
`item_FORM_SUBMITTED_FAILURE`

I wonder if we should do it in another PR? as I don't know for sure where is it used except for `bookmarks`

File: webpack/stories/docs/api-middleware.stories.mdx
Comment (Ron-Lavi) [reply webpack/stories/docs/api-middleware.stories.mdx]: I think from one of the rebases.. fixed, thanks !

File: webpack/assets/javascripts/react_app/redux/actions/common/forms.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/redux/actions/common/forms.js:67]: I was thinking the same thing. I'd rather do this in a separate PR, [forman_templates](https://github.com/theforeman/foreman_templates/blob/master/webpack/components/TemplateSyncResult/TemplateSyncResultReducer.js#L22) will be affected as well.

**Comment by johnpmitsch** [comment]
I'm using this while creating the details page of the new CV page and its working well so far!

One pattern that I think we will see a lot is updating/creating/deleting something and creating a toast notification when it is finished successfully or if there was an error. This can be done as-is, but it gets a bit cumbersome.

Here is an example of just a PUT call, with no notifications or feedback:
```javascript
export const updateContentView = (cvId, params) => put({
      type: API_OPERATIONS.PUT,
      key: `${CONTENT_VIEWS_KEY}_${cvId}`,
      url: api.getApiUrl(`/content_views/${cvId}`),
      params,
});
```

This is pretty simple and works well. 

It gets complicated when you want a notification on success or error. toasts need to be dispatched as an action, so this means you need access to the dispatch and use a wrapper function where the dispatch is passed in and then dispatch the notifications with that.
```javascript

export const updateContentView = (cvId, params) => async (dispatch) => {
  return dispatch(
    put({
      type: API_OPERATIONS.PUT,
      key: `${CONTENT_VIEWS_KEY}_${cvId}`,
      url: api.getApiUrl(`/content_views/${cvId}`),
      params,
      handleSuccess: () => dispatch(
        addToast({
          type: 'success',
          message: __(' Content View updated.'),
        })
      ),      
      handleError: (error) => dispatch(apiError(null, error)), // helper function for errors
    })
  );
};
```
This does work fine, but I could see something like this being beneficial:

```javascript
export const updateContentView = (cvId, params) => put({
  type: API_OPERATIONS.PUT,
  key: `${CONTENT_VIEWS_KEY}_${cvId}`,
  url: api.getApiUrl(`/content_views/${cvId}`),
  params,
  successNotificationText: "Content View Updated",
  errorNotificationText: "Something went wrong!",
})
```

Here the dispatch would not need to be passed in. We can dispatch the success notification in the middleware if `successNotificationText` is present. If its not present, no notification is dispatched.

For the error notification, you could even just make it as a boolean e.g. `errorNotification: true` and use [something like this](https://github.com/Katello/katello/blob/master/webpack/move_to_foreman/common/helpers.js#L26-L38) that gets the error from the response and creates a notification with that.

I imagine the POST/PUT/DELETE calls will involve notifications more often than not, so IMHO its worth considering adding an easy way to create them here.

**Comment by Ron-Lavi** [comment]
@johnpmitsch NIce! I was thinking about the toasts also!
I like any shortcut we can provide to make life easier for developers,
especially for common stuff :)

what I am suggesting would be in a follow-up PR,
maybe we can add a `toast` key to the action, e.g:
```js
export const updateContentView = (cvId, params) => put({
  key,
  url,
  params,
  toast: {
    success: "Content View Updated",
    error: "Something went wrong!",
  }
})
```

**Comment by johnpmitsch** [comment]
> @johnpmitsch NIce! I was thinking about the toasts also!
> I like any shortcut we can provide to make life easier for developers,
> especially for common stuff :)
> 
> what I am suggesting would be in a follow-up PR,
> maybe we can add a `toast` key to the action, e.g:
> 
> ```js
> export const updateContentView = (cvId, params) => put({
>   key,
>   url,
>   params,
>   toast: {
>     success: "Content View Updated",
>     error: "Something went wrong!",
>   }
> })
> ```

@laviro Great! A separate PR is fine unless you feel its an easy addition here.

I think the error handling will need to support showing the actual error message, which will only be known when the action is called and the API sends back an error. I'm thinking something like this:
```javascript
// Automatically handle error by showing notification with returned error message from API. 
// This would be the default when `toast` is used and no `error` key is specified. `success` key is required.
toast: {
   success: "Content View Updated",
}

// Override error message to a fixed string, use this for error notification.
toast: {
  success: "Content View Updated",
  error: "Something went wrong"
}

// Create a custom notification text that has access to the response or error object by passing in a function.
// Note this is just to create the text, not to actually create the toast.
// That would still be done by the middleware, so the dispatch does not need to be passed in.
toast: {
  success: (response) => myCustomSuccessMessage(response), // returns a string dynamically created with the response object
  error: (error) => myCustomErrorMessage(error) // return a string dynamically created with the error object
}
```

In the first example, the error message is shown by default as a notification, so if the API returns "name is too long", an error notification will be created with this text. The second uses fixed strings as messages, and the third would allow custom text in the notification that has access to the response and error message. I can see the custom message useful if you want to use the name or label in a success message or some sort of complex error handling logic.

**Comment by Ron-Lavi** [comment]
> @laviro Great! A separate PR is fine unless you feel its an easy addition here.

thanks! yeah I will create a separate PR for the toasts shortcut feature,
as this small PR is open from December 2019

**Comment by Ron-Lavi** [comment]
rebased

**Review by johnpmitsch — DISMISSED** [review]
I can say this has been working fine in https://github.com/Katello/katello/pull/8808 and it seems others have reviewed the code thoroughly.

@laviro Are there any concerns left to address?

**Comment by Ron-Lavi** [comment]
Thanks @johnpmitsch for the approve,
I just rebased to get back the GH actions to test with plugins,

> @laviro Are there any concerns left to address?

I think I answered it all :+1: lets get this in :)

**Comment by jeremylenz** [comment]
@amirfefer any other concerns?  We've [started using it already](https://github.com/Katello/katello/pull/8808) so would be great to get it merged :)

**Review by ezr-ondrej — APPROVED** [review]
Thanks @laviro for doing this.
Thanks @MariaAga, @johnpmitsch, @jeremylenz, @xprazak2 and @amirfefer for reviews!
Given this was thoroughly reviewed by many people and all the requested changes have been incorporated, I've just checked this doesn't break current api requests.

**Comment by Ron-Lavi** [comment]
Yay!! thanks @ezr-ondrej and thanks everyone for the review :):)

**Pull request merged**
